OrderedLockTest.testComplex fails on alll platforms since I20220904-1800

OrderedLockTest.testComplex fails on alll platforms since I20220904-1800

Regression from https://github.com/eclipse-platform/eclipse.platform/pull/127

For example in https://download.eclipse.org/eclipse/downloads/drops4/I20220904-1800/testresults/html/org.eclipse.core.tests.runtime_ep426I-unit-cen64-gtk3-java11_linux.gtk.x86_64_11.html

```
testComplex(org.eclipse.core.tests.runtime.jobs.OrderedLockTest): failures 2 out of 10 tries. Last cause:java.lang.IllegalStateException: thread did not end

java.lang.AssertionError: testComplex(org.eclipse.core.tests.runtime.jobs.OrderedLockTest): failures 2 out of 10 tries. Last cause:java.lang.IllegalStateException: thread did not end
at org.eclipse.core.tests.runtime.jobs.RetryTestRule$1.evaluate(RetryTestRule.java:50)
```

--- a/runtime/bundles/org.eclipse.core.jobs/META-INF/MANIFEST.MF
+++ b/runtime/bundles/org.eclipse.core.jobs/META-INF/MANIFEST.MF
@@ -6,6 +6,7 @@ Bundle-Version: 3.13.200.qualifier
 Bundle-Vendor: %providerName
 Bundle-Localization: plugin
 Export-Package: org.eclipse.core.internal.jobs;x-internal:=true,
+ org.eclipse.core.internal.jobs;x-friends:="org.eclipse.core.tests.runtime.jobs",
  org.eclipse.core.runtime.jobs
 Bundle-Activator: org.eclipse.core.internal.jobs.JobActivator
 Require-Bundle: org.eclipse.equinox.common;bundle-version="[3.8.0,4.0.0)"

--- a/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/DeadlockDetector.java
+++ b/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/DeadlockDetector.java
@@ -68,7 +68,7 @@ import org.eclipse.core.runtime.jobs.ISchedulingRule;
  * Therefore, in order to make sure that no deadlock is possible,
  * the deadlock will still be resolved at this point.
  */
-class DeadlockDetector {
+public class DeadlockDetector {
 	private static int NO_STATE = 0;
 	//state variables in the graph
 	private static int WAITING_FOR_LOCK = -1;
@@ -82,6 +82,7 @@ class DeadlockDetector {
 	private final ArrayList<Thread> lockThreads = new ArrayList<>();
 	//whether the graph needs to be resized
 	private boolean resize = false;
+	private static volatile boolean noDeadlockReport;
 
 	/**
 	 * Recursively check if any of the threads that prevent the current thread from running
@@ -597,9 +598,24 @@ class DeadlockDetector {
 	}
 
 	/**
+	 * For tests that are supposed to create deadlocks.
+	 */
+	public static void runSilent(Runnable runnable) {
+		DeadlockDetector.noDeadlockReport = true;
+		try {
+			runnable.run();
+		} finally {
+			DeadlockDetector.noDeadlockReport = false;
+		}
+	}
+
+	/**
 	 * Adds a 'deadlock detected' message to the log with a stack trace.
 	 */
 	private void reportDeadlock(Deadlock deadlock) {
+		if (noDeadlockReport) {
+			return;
+		}
 		String msg = "Deadlock detected. All locks owned by thread " + deadlock.getCandidate().getName() + " will be suspended."; //$NON-NLS-1$ //$NON-NLS-2$
 		MultiStatus main = new MultiStatus(JobManager.PI_JOBS, JobManager.PLUGIN_ERROR, msg, new IllegalStateException());
 		Thread[] threads = deadlock.getThreads();

--- a/runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/OrderedLockTest.java
+++ b/runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/OrderedLockTest.java
@@ -18,8 +18,7 @@ import static org.junit.Assert.assertTrue;
 import java.util.*;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.stream.Collectors;
-import org.eclipse.core.internal.jobs.LockManager;
-import org.eclipse.core.internal.jobs.OrderedLock;
+import org.eclipse.core.internal.jobs.*;
 import org.eclipse.core.runtime.jobs.ILock;
 import org.eclipse.core.runtime.jobs.LockListener;
 import org.eclipse.core.tests.harness.TestBarrier2;
@@ -46,32 +45,36 @@ public class OrderedLockTest {
 
 	@Test
 	public void testComplex() {
-		ArrayList<LockAcquiringRunnable> allRunnables = new ArrayList<>();
-		LockManager manager = new LockManager();
-		OrderedLock lock1 = manager.newLock();
-		OrderedLock lock2 = manager.newLock();
-		OrderedLock lock3 = manager.newLock();
-		createRunnables(new ILock[] {lock1, lock2, lock3}, 5, allRunnables);
-		createRunnables(new ILock[] {lock3, lock2, lock1}, 5, allRunnables);
-		createRunnables(new ILock[] {lock1, lock3, lock2}, 5, allRunnables);
-		createRunnables(new ILock[] {lock2, lock3, lock1}, 5, allRunnables);
-		execute(allRunnables);
-		//the underlying array has to be empty
-		assertTrue("Locks not removed from graph.", manager.isEmpty());
+		DeadlockDetector.runSilent(() -> {
+			ArrayList<LockAcquiringRunnable> allRunnables = new ArrayList<>();
+			LockManager manager = new LockManager();
+			OrderedLock lock1 = manager.newLock();
+			OrderedLock lock2 = manager.newLock();
+			OrderedLock lock3 = manager.newLock();
+			createRunnables(new ILock[] { lock1, lock2, lock3 }, 5, allRunnables);
+			createRunnables(new ILock[] { lock3, lock2, lock1 }, 5, allRunnables);
+			createRunnables(new ILock[] { lock1, lock3, lock2 }, 5, allRunnables);
+			createRunnables(new ILock[] { lock2, lock3, lock1 }, 5, allRunnables);
+			execute(allRunnables);
+			// the underlying array has to be empty
+			assertTrue("Locks not removed from graph.", manager.isEmpty());
+		});
 	}
 
 	@Test
 	public void testSimple() {
-		ArrayList<LockAcquiringRunnable> allRunnables = new ArrayList<>();
-		LockManager manager = new LockManager();
-		OrderedLock lock1 = manager.newLock();
-		OrderedLock lock2 = manager.newLock();
-		OrderedLock lock3 = manager.newLock();
-		createRunnables(new ILock[] {lock1, lock2, lock3}, 1, allRunnables);
-		createRunnables(new ILock[] {lock3, lock2, lock1}, 1, allRunnables);
-		execute(allRunnables);
-		//the underlying array has to be empty
-		assertTrue("Locks not removed from graph.", manager.isEmpty());
+		DeadlockDetector.runSilent(() -> {
+			ArrayList<LockAcquiringRunnable> allRunnables = new ArrayList<>();
+			LockManager manager = new LockManager();
+			OrderedLock lock1 = manager.newLock();
+			OrderedLock lock2 = manager.newLock();
+			OrderedLock lock3 = manager.newLock();
+			createRunnables(new ILock[] { lock1, lock2, lock3 }, 1, allRunnables);
+			createRunnables(new ILock[] { lock3, lock2, lock1 }, 1, allRunnables);
+			execute(allRunnables);
+			// the underlying array has to be empty
+			assertTrue("Locks not removed from graph.", manager.isEmpty());
+		});
 	}
 
 	@Test
@@ -288,9 +291,13 @@ public class OrderedLockTest {
 			thread.start();
 		}
 		randomOrder.waitForEnd();
+		long joinMillis = 500;
 		for (Thread thread : threads) {
 			try {
-				thread.join(500);
+				long n0 = System.nanoTime();
+				thread.join(joinMillis);
+				long n1 = System.nanoTime();
+				joinMillis = Math.max(joinMillis - (n1 - n0) / 1000_000, 1);
 				if (thread.isAlive()) {
 					throw new IllegalStateException("thread did not end");
 
@@ -300,4 +307,5 @@ public class OrderedLockTest {
 			}
 		}
 	}
+
 }

